libjcat/jcat-bt-util.c

```diff
@@ -144,3 +146,405 @@ jcat_bt_generate_key_pair(const gchar *keyname,
+#ifndef ENABLE_ED25519
+	g_set_error_literal(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    __FUNCTION__ " requires Ed25519 support but it is not enabled");
+	return FALSE;
+#else
+
+	/* Creates a signed note in the Go sumdb format. */
+	GByteArray *buf = NULL;
+	g_autoptr(GByteArray) raw_signature = g_byte_array_set_size(g_byte_array_new(), 64);
+	g_autofree gchar *encoded_signature = NULL;
+	g_autoptr(GString) formatted_signature = g_string_new(NULL);
+
+	g_return_val_if_fail(text != NULL, NULL);
+	g_return_val_if_fail(parsed_private_key != NULL, NULL);
+	g_return_val_if_fail(parsed_key_name != NULL, NULL);
+	g_return_val_if_fail(parsed_key_hash != NULL, NULL);
+
+	/* Require the text to end with a newline */
+	if (text->len == 0 || text->str[text->len - 1] != '\n') {
+		g_set_error(error,
+			    G_IO_ERROR,
+			    G_IO_ERROR_FAILED,
+			    "Cannot sign: text must end with a newline");
+		return NULL;
+	}
+
+	ed25519_sha512_sign(parsed_public_key->data,
+			    parsed_private_key->data,
+			    text->len,
+			    (const guint8 *)text->str,
+			    raw_signature->data);
+
+	/* Add the key hash prior to the signature */
+	g_byte_array_prepend(raw_signature, (const guint8 *)parsed_key_hash, 4);
+
+	encoded_signature = g_base64_encode(raw_signature->data, raw_signature->len);
+
+	/* Add text and an extra new line (need two consecutive newlines) */
+	buf = g_byte_array_new();
+	g_byte_array_append(buf, (const guint8 *)text->str, text->len);
+	g_byte_array_append(buf, (const guint8 *)"\n", 1);
+
+	/* Format the signature */
+	g_string_printf(formatted_signature,
+			"\xe2\x80\x94 %s %s\n",
+			parsed_key_name,
+			encoded_signature);
+	g_byte_array_append(buf,
+			    (const guint8 *)formatted_signature->str,
+			    formatted_signature->len);
+	return g_byte_array_free_to_bytes(buf);
+#endif
+}
+
+/*
+ The on-disk structure is:
+
+	<rootDir>/leaves/aa/bb/cc/ddeeff...
+	<rootDir>/leaves/pending/aabbccddeeff...
+	<rootDir>/seq/aa/bb/cc/ddeeff...
+	<rootDir>/tile/<level>/aa/bb/ccddee...
+	<rootDir>/checkpoint
+ */
+
+static gboolean
+create_storage(GFile *storage_dir, GError **error)
+{
+	gboolean ret;
+	g_autoptr(GFile) seq_dir = NULL;
+	g_autoptr(GFile) tile_dir = NULL;
+	g_autoptr(GFile) leaves_dir = NULL;
+	g_autoptr(GFile) pending_dir = NULL;
+
+	ret = g_file_make_directory_with_parents(storage_dir, NULL, error);
+	if (!ret)
+		return FALSE;
+
+	/* Create necessary subdirectories */
+	seq_dir = g_file_get_child(storage_dir, "seq");
+	ret = g_file_make_directory(seq_dir, NULL, error);
+	if (!ret)
+		return FALSE;
+
+	tile_dir = g_file_get_child(storage_dir, "tile");
+	ret = g_file_make_directory(tile_dir, NULL, error);
+	if (!ret)
+		return FALSE;
+
+	leaves_dir = g_file_get_child(storage_dir, "leaves");
+	ret = g_file_make_directory(leaves_dir, NULL, error);
+	if (!ret)
+		return FALSE;
+
+	pending_dir = g_file_get_child(leaves_dir, "pending");
+	ret = g_file_make_directory(pending_dir, NULL, error);
+	if (!ret)
+		return FALSE;
```

Comment:
i don't think using `ret` is useful here.

Reply:
Done.